clients/src/main/java/org/apache/kafka/common/metrics/KafkaMetricsContext.java

```diff
@@ -0,0 +1,64 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.common.metrics;
+
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * A implementation of MetricsContext, it encapsulates required metrics context properties for Kafka
+ */
+public class KafkaMetricsContext implements MetricsContext {
+    public static final String METRICS_CONTEXT_PREFIX = "metrics.context.";
+
+    /**
+     * Client or Service's metadata map.
+     */
+    private Map<String, String> metadata = new HashMap<>();
+
+    public KafkaMetricsContext() {
```

Comment:
do we need this constructor? In the KIP we mention that the namespace prefix will always be provided.

Reply:
This is added to support this constructor: https://github.com/apache/kafka/blob/9bc96d54f8953d190a1fb6478a0656f049ee3b32/clients/src/main/java/org/apache/kafka/common/metrics/Metrics.java#L131

It is used by other constructors which simply pass in empty reporter list. I can remove this constructor and pass in an empty string when no prefix is specified:
        this(defaultConfig, reporters, time, enableExpiration, new KafkaMetricsContext(""));

Reply:
in which cases do we not use a jmx prefix?

Reply:
In some of Metrics constructor,  there is no metrics reporter, they are used:
https://github.com/xiaodongdu/kafka/blob/e8938501520273a42353cddcb30b5a7acb2a0f27/connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorTaskConfig.java#L57

https://github.com/xiaodongdu/kafka/blob/e8938501520273a42353cddcb30b5a7acb2a0f27/clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java#L510

https://github.com/xiaodongdu/kafka/blob/e8938501520273a42353cddcb30b5a7acb2a0f27/jmh-benchmarks/src/main/java/org/apache/kafka/jmh/metadata/MetadataRequestBenchmark.java#L98
https://github.com/xiaodongdu/kafka/blob/e8938501520273a42353cddcb30b5a7acb2a0f27/jmh-benchmarks/src/main/java/org/apache/kafka/jmh/fetcher/ReplicaFetcherThreadBenchmark.java#L246

https://github.com/xiaodongdu/kafka/blob/e8938501520273a42353cddcb30b5a7acb2a0f27/tools/src/main/java/org/apache/kafka/trogdor/workload/ConnectionStressWorker.java#L165

Others are from tests.

Reply:
The one from KafkaAdminClient, is used for unit test:
https://github.com/xiaodongdu/kafka/blob/e8938501520273a42353cddcb30b5a7acb2a0f27/clients/src/test/java/org/apache/kafka/clients/admin/AdminClientUnitTestEnv.java#L95

Reply:
The other one is from MirrorTaskConfig, which is used in MirrorSourceTask and MirrorCheckPointTask. 
The rest are from benchmark.